search.php

```diff
@@ -61,130 +72,178 @@ function showSearchResults($svnrep, $path, $searchstring, $rev, $peg, $listing,
+		$listvar['rowparity'] = $index % 2;
+
+		if ($isDir) 
+		{
+			$listvar['filetype'] = ($openDir) ? 'diropen' : 'dir';
+			$openDir = true;
+		}
+		else 
+		{
+			$listvar['filetype'] = strtolower(strrchr($file, '.'));
+			$openDir = false;
+		}
+
+		$listvar['isDir'] = $isDir;
+		$listvar['openDir'] = $openDir;
+		$listvar['path'] = $path.$file;
+
+		$tempelements = explode('/',$file);
+
+		if ($tempelements[count($tempelements)-1] === "")
+		{
+			$lastindexfile = count($tempelements)-1 - 1;
+			$listvar['node'] = $lastindexfile; // t-node
+			$listvar['level'] = ($treeview) ? $lastindexfile : 0;
+			$listvar['filename'] = $tempelements[$lastindexfile];
+		}
+		else
+		{
+			$lastindexfile = count($tempelements)-1;
+			$listvar['node'] = $lastindexfile; // t-node
+			$listvar['level'] = ($treeview) ? $lastindexfile : 0;
+			$listvar['filename'] = $tempelements[$lastindexfile];
+		}
+
+		for ($j=1;$j<=$lastindexfile;$j++)
+		{
+			$listvar['last_i_node'][$j] = false;
+		}
+
+		if ($isDir) 
+		{
+			$listvar['fileurl'] = urlForPath($path.$file, $passRevString);
+		}
+		else
+		{
+			$listvar['fileurl'] = urlForPath($path.$file, createDifferentRevAndPegString($passrev, $peg));
+		}
+
+		$listvar['filelink'] = '<a href="'.$listvar['fileurl'].'">'.$listvar['filename'].'</a>';
+
+		if ($isDir) 
+		{
+			$listvar['logurl'] = $config->getURL($rep, $path.$file, 'log').$isDirString.$passRevString;
+		}
+		else
+		{
+			$listvar['logurl'] = $config->getURL($rep, $path.$file, 'log').$isDirString.createDifferentRevAndPegString($passrev, $peg);
+		}
+
+		if ($treeview)
+		{
+			$listvar['compare_box'] = '<input type="checkbox" name="compare[]" value="'.escape($path.$file).'@'.$passrev.'" onclick="enforceOnlyTwoChecked(this)" />';
+		}
+
+		if ($config->showLastModInListing()) 
+		{
+			$listvar['committime'] = $entry->committime;
+			$listvar['revision'] = $entry->rev;
+			$listvar['author'] = $entry->author;
+			$listvar['age'] = $entry->age;
+			$listvar['date'] = $entry->date;
+			$listvar['revurl'] = $config->getURL($rep, $path.$file, 'revision').$isDirString.createRevAndPegString($entry->rev, $peg ? $peg : $rev);
+		}
+
+		if ($rep->isDownloadAllowed($path.$file))
+		{
+			$downloadurl = $config->getURL($rep, $path.$file, 'dl').$isDirString.$downloadRevAndPeg;
+
+			if ($isDir) 
+			{
+				$listvar['downloadurl'] = $downloadurl;
+				$listvar['downloadplainurl'] = '';
+			}
+			else
+			{
+				$listvar['downloadplainurl'] = $downloadurl;
+				$listvar['downloadurl'] = '';
 			}
 		}
+		else
+		{
+			$listvar['downloadplainurl'] = '';
+			$listvar['downloadurl'] = '';
+		}
+
+		if ($rep->isRssEnabled())
+		{
+			// RSS should always point to the latest revision, so don't include rev
+			$listvar['rssurl'] = $config->getURL($rep, $path.$file, 'rss').$isDirString.createRevAndPegString('', $peg);
+		}
+
+		$loop++;
+		$index++;
+		$last_index = $index;
 	}
 
 	return $listing;
 }
+
 // Make sure that we have a repository
-if ($rep) {
+if ($rep)
```

Comment:
You should change this as well. This is unnecessary bad style in many files currently:

```
if (!$rep)
{
    http_response_code(404);
}

[...]
```

Makes things much more readable.

Reply:
But this is in an if else case.
```
if ($rep) { --- }
else { http_response_code(404);}
renderTemplate('directory')
```
So if we do this the other section will still be evaluated when it should not. As http_response_code i think only updates the header doesnt exits the code.

Reply:
You are right, I didn't make things explicit I had in mind. My suggestion should really have been the following:

```
if (!$rep)
{
    http_response_code(404);
    renderTemplate('directory');
    exit(0);
}

[...]
renderTemplate('directory');
```

While that duplicates the call to `renderTemplate`, aborting early in my opinion is better to read and maintain. One could even make things easier with something like the following, putting the call to `exit` into the function:

```
function renderTemplate($view, $andExit = false) {
```

Or something like that, handling the whole error case in one call:

```
function renderTemplateNoRep($view)
{
    http_response_code(404);
    renderTemplate($view);
    exit(0);
}
```

That combination of checking for an available repo and print `404` otherwise happens pretty often. At some point the template for that will most likely change as well, because that's an error which would need to be printed specially to the user. Doesn't make too much sense to print a template designed with a repo being available in mind when it's actually not. Introducing such a function would make it easy to find all those places by name.

Reply:
I have made an exit first implementation. Though i am thinking should we wrap the rest of the code into a function? Though not sure would it help in some way.

Reply:
Functions often help, because they can be documented and stuff like that. But just putting all available code into one without looking if that makes sense as one function, if it makes sense to document the behaviour of that function etc. doesn't help too much. You just would move the code from one place to another, but it would still be a "do everything"-stuff. That doesn't make too much sense, it would only help if you would break the existing code into pieces worth being maintained as individual function per piece, with individual, concrete docs per function etc.

Look at the example I brought with `renderTemplateNoRepo`:

```
http_response_code(404);
renderTemplate('directory');
exit(0);
```

Instead of copying those three lines over and over again in different files, they need to be outsourced into one function used by all files. That function can be documented. maintained at one place only etc. This is different to simply putting available code in one function just to have it as one function, functions need to serve one concrete purpose like rendering things or processing things or retrieving things or checking things or ...

The bottom line is, keep the available large code as is, because putting it in a function wouldn't improve things a lot. If at all, introduce `renderTemplateNoRepo` instead, because the changes you made now need to be made at some point in future in many other files as well.

Reply:
I made a function in the utils and pushed it in. This can be used in all other templates as well. we can create a backlog item to fix other files with this function as well. 

For the 404 error handle i totally agree with you. I am still not sure why we are printing the entire template that looks weird as seen in image in #126 

> At some point the template for that will most likely change as well, because that's an error which would need to be printed specially to the user. Doesn't make too much sense to print a template designed with a repo being available in mind when it's actually not.

We should discuss the change in template for error 404 in #126.